samples/Hardware.Esp32.Rmt/InfraredRemoteReceiver/InfraredRemoteReceiverSample/Decoder.cs

```diff
@@ -0,0 +1,157 @@
+﻿//
+// Copyright (c) .NET Foundation and Contributors
+// See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Text;
+using nanoFramework.Hardware.Esp32.Rmt;
+
+namespace InfraredRemote
+{
+    public abstract class Decoder
+    {
+        /// <summary>
+        /// Defines constructor of <see cref="Decoder"/>.</typeparam>
+        /// </summary>
+        /// <param name="signalLengthTolerance">Tolerance of signal length represented as fraction value. </param>
+        protected Decoder(double signalLengthTolerance)
+        {
+            SignalLengthTolerance = signalLengthTolerance;
+        }
+        protected SignalData lastData = null;
+        protected abstract int PulseLengthInMicroseconds { get; }
+
+        protected abstract int HeaderMark { get; }
+        protected abstract int HeaderSpace { get; }
+        protected abstract int RepeatSpace { get; }
+
+        protected abstract int ZeroSpace { get; }
+
+        protected abstract int OneSpace { get; }
+
+        protected abstract Protocol Protocol { get; }
+        protected double SignalLengthTolerance { get; private set; } = 0.2;
+
+        protected abstract int SignalLength { get; }
+
+        protected abstract bool UseLessSignificantBitFirst { get; }
+
+        protected abstract int AddressBits { get; }
+        protected abstract int CommandBits { get; }
+
+        /// <summary>
+        /// Decodes Rmt signal into SignalData.
+        /// </summary>
+        /// <param name="receivedSignal">Array representing decoded signal.</param>
+        /// <returns>SignalData object.</returns>
+        public virtual SignalData Decode(RmtCommand[] receivedSignal)
+        {
+            SignalData signalData = new SignalData();
+
+            var firstPulse = receivedSignal[0];
+            var lastPulse = receivedSignal[receivedSignal.Length - 1];
+            bool isHeaderMark = Match(firstPulse.Duration0, HeaderMark);
+            bool isSpaceMark = Match(firstPulse.Duration1, HeaderSpace);
+            bool isEndOfTransmission = Match(lastPulse.Duration0, PulseLengthInMicroseconds) && !lastPulse.Level0 && lastPulse.Duration1 == 0 &&
+                                       lastPulse.Level1;
+            bool isRepeat = receivedSignal.Length == 2 && Match(firstPulse.Duration0, HeaderMark) && !firstPulse.Level0 && Match(firstPulse.Duration1, RepeatSpace) && firstPulse.Level1;
+
+            var message = ExtractRawPayload(receivedSignal);
```

Comment:
I think the call to `ExtractRawPayload` call needs to happen after the `if (isRepeat)` and inside the `then` condition of the `if(isHeaderMark && ....)` check. Less CPU cycles used this way.